cmd/recovereth/main.go

```diff
@@ -0,0 +1,396 @@
+				fmt.Printf("Recovery successfull!\n")
+				fmt.Printf("Recovered password: \n------------\n%v\nHex: 0x%x\n------------\n", pw, pw)
+				fmt.Printf("Please report to the go-ethereum repository "+
+					"that you were able to recover a wallet with prefix 0x%x and suffix 0x%x.", prefix, suffix)
+				return nil
+			}
+		}
+	}
+	return errors.New("Exhausted")
+}
+func testPasswordCrops(keyjson, auth string) error {
+	k := new(encryptedKeyJSONV3)
+	if err := json.Unmarshal([]byte(keyjson), k); err != nil {
+		log.Fatal(err)
+	}
+	addr := common.HexToAddress(k.Address)
+	for i:= 0 ; i < len(auth); i++ {
+		pw := auth[:i]
+		keyBytes, err := decryptKeyV3x(k.Crypto, pw)
+		fmt.Printf("auth: %x\n", pw)
+		if err != nil {
+			return err
+		}
+		key := crypto.ToECDSAUnsafe(keyBytes)
+		if err != nil {
+			return err
+		}
+		recovered := crypto.PubkeyToAddress(key.PublicKey)
+		fmt.Printf("found address %s (looking for %s)\n", recovered.Hex(), addr.Hex())
+		if recovered == addr {
+			fmt.Printf("Recovery successfull!\n")
+			fmt.Printf("Recovered password: \n------------\n%v\nHex: 0x%x\n------------\n", pw, pw)
+			fmt.Printf("Please report to the go-ethereum repository "+
+				"that you were able to recover a wallet by cropping %d-length password to length %d ", len(auth), len(pw))
+			return nil
+		}
+	}
+	return errors.New("Exhausted")
+}
+
+func testRecoveryBitFlip(keyjson, auth string, i uint) error {
+	// i = 0 starts on the Ciphertext
+	// i = 256 starts on the IV
+	// i = 384 starts on the SALT
+	for ; ; i++ {
+		k := new(encryptedKeyJSONV3)
+
+		if err := json.Unmarshal([]byte(keyjson), k); err != nil {
+			return err
+		}
+		addr := common.HexToAddress(k.Address)
+
+		k2, err := flipBit(k.Crypto, i)
+		if err != nil {
+			return err
+		}
+
+		fmt.Printf("C  %s -> %s \n", k.Crypto.CipherText, k2.CipherText)
+		fmt.Printf("IV %s -> %s \n", k.Crypto.CipherParams.IV, k2.CipherParams.IV)
+		fmt.Printf("Sa %s -> %s \n", k.Crypto.KDFParams["salt"].(string), k2.KDFParams["salt"].(string))
+
+		keyBytes, err := decryptKeyV3x(k2, auth)
+		if err != nil {
+			return err
+		}
+		key := crypto.ToECDSAUnsafe(keyBytes)
+		if err != nil {
+			return err
+		}
+		recovered := crypto.PubkeyToAddress(key.PublicKey)
+		fmt.Printf("found address %s (looking for %s)\n", recovered.Hex(), addr.Hex())
+		if recovered == addr {
+			fmt.Printf("Recovery successfull!\n")
+			json.Marshal(k2)
+			json, _ := json.MarshalIndent(k2, "", "    ")
+			fmt.Printf("Recovered json-file: \n------------\n%v\n------------\n", string(json))
+			fmt.Printf("Please report to the go-ethereum repository "+
+				"that you were able to recover a wallet by 'bitflip'-manipulation at bit %d.", i)
+			return nil
+		}
+	}
+}
+
+func testDerivedKeyBitflip(keyjson, auth string) error {
+	k := new(encryptedKeyJSONV3)
+	if err := json.Unmarshal([]byte(keyjson), k); err != nil {
+		log.Fatal(err)
+	}
+	addr := common.HexToAddress(k.Address)
+	iv, _ := hex.DecodeString(k.Crypto.CipherParams.IV)
```

Comment:
I think we need to add error checking here
```go
	iv, err := hex.DecodeString(k.Crypto.CipherParams.IV)
	if err != nil {
		log.Fatalf("Failed to decode IV: %v", err)
	}
```

Reply:
Nah, that's kind of the idea. We don't exit "defensively" on errors, we forge ahead. The only measure of success is whether we succeed in recovering the key. 

E.g. if a keystore has a MAC, and the MAC is broken because the MAC-field is corrupt. THen we want to skip that, recover they key anyway. Not fail reading the key just because some metadata is broken.